### PR TITLE
CLI-84: Implemented file references

### PR DESCRIPTION
**Pull request by dokov**

### CODE REVIEW

**Comment by dokov** [comment]
Agreed. Resolving logic moved to ConfigFileProcessor.

File: lib/ConfigFileProcessor.js [hunk 1]
```diff
@@ -75,6 +80,79 @@ class ConfigFileProcessor {
...
     });
   }
+
+  /**
+   * Reads a JSON configuration file and returns a JavaScript object. Does all the necessary processing and resolves all referenced files.
+   * @param {string} filePath
+   */
+  static readJSONConfigFile(file, cb) {
+    ConfigFileProcessor._readJSONConfigFile(file, 'root', (err, config) => {
+      logger.debug('Final configuration: ' + JSON.stringify(config, null, 4));
```
Comment (DennyGD) [inline lib/ConfigFileProcessor.js]: What if there is an error here?

File: lib/ConfigFileProcessor.js [hunk 2]
```diff
@@ -75,6 +80,79 @@ class ConfigFileProcessor {
...
+      }
+    ], cb);
+  }
+
+  static _processConfigFileValue(obj, baseFolder, fieldPath, done) {
+    if (Array.isArray(obj) || lodashIsObject(obj)) {
+      async.forEachOfSeries(
+        obj,
+        (value, key, callback) => {
+          // console.log(key + '    ' + JSON.stringify(value));
```
Comment (DennyGD) [inline lib/ConfigFileProcessor.js]: Forgotten, I guess

File: lib/ConfigFileProcessor.js [hunk 3]
```diff
@@ -75,6 +80,79 @@ class ConfigFileProcessor {
...
+          logger.debug('Referenced files resolved for %s', chalk.cyan(file));
+
+          return next(null, processedValue);
+        });
+      }
+    ], cb);
+  }
+
+  static _processConfigFileValue(obj, baseFolder, fieldPath, done) {
+    if (Array.isArray(obj) || lodashIsObject(obj)) {
```
Comment (DennyGD) [inline lib/ConfigFileProcessor.js]: I think that `isObject` from lodash will return true for an array, although I haven't tested it

File: lib/ConfigFileProcessor.js [hunk 4]
```diff
@@ -13,11 +13,16 @@
  * contents is a violation of applicable laws.
  */
 
+const fs = require('fs');
 const async = require('async');
-
-const { ConfigType } = require('./Constants');
+const path = require('path');
+const logger = require('./logger.js');
```
Comment (DennyGD) [inline lib/ConfigFileProcessor.js]: For better or worse, I've always envisioned that the `logger` module will be used through the `CLIManager`'s `log` method. 'Cause this is where the `logger` is 'initialized' depending on the options passed to the CLI by the user. And now I see that even there's additional logic in this `log` method. So I think that either some refactoring should occur or logging here should happen through the `log` method.

File: lib/ConfigFileProcessor.js (hunk 1 above)
Comment (dokov) [reply lib/ConfigFileProcessor.js]: Makes sense, we wouldn't want this to print if there is an error, I will put it in an "if".

File: lib/ConfigFileProcessor.js (hunk 2 above)
Comment (dokov) [reply lib/ConfigFileProcessor.js]: Not really, but will remove it :)

File: lib/ConfigFileProcessor.js (hunk 3 above)
Comment (dokov) [reply lib/ConfigFileProcessor.js]: Good catch. Will change it to isPlainObject.

File: lib/ConfigFileProcessor.js (hunk 4 above)
Comment (dokov) [reply lib/ConfigFileProcessor.js]: Hmmm, I copied this from Utils.js where the logger is used directly. In general it is a good practice for the logger to be a signleton, e.g. have a configured instance that is accessible from anywhere.

Comment (DennyGD) [reply lib/ConfigFileProcessor.js]: I'm not opposing the idea 😃 
The usage in Utils was a leftover meant to be changed. I don't think we should have logging in Utils functions, anyways.

**Pull request merged**